core/src/main/java/org/apache/carbondata/core/datastore/SegmentTaskIndexStore.java

```diff
@@ -338,14 +339,28 @@ private synchronized Object addAndGetSegmentLock(String segmentId) {
    * @throws IOException
    */
   private AbstractIndex loadBlocks(TaskBucketHolder taskBucketHolder,
-      List<TableBlockInfo> tableBlockInfoList, AbsoluteTableIdentifier tableIdentifier)
+      List<TableBlockInfo> tableBlockInfoList, AbsoluteTableIdentifier tableIdentifier,
+      TableSegmentUniqueIdentifier tableSegmentUniqueIdentifier)
       throws IOException {
     // all the block of one task id will be loaded together
     // so creating a list which will have all the data file meta data to of one task
     List<DataFileFooter> footerList = CarbonUtil
         .readCarbonIndexFile(taskBucketHolder.taskNo, taskBucketHolder.bucketNumber,
             tableBlockInfoList, tableIdentifier);
-    AbstractIndex segment = new SegmentTaskIndex();
+
+    if (null == tableSegmentUniqueIdentifier.getSegmentProperties()) {
+      // create a metadata details
+      // this will be useful in query handling
+      // all the data file metadata will have common segment properties we
+      // can use first one to get create the segment properties
+      SegmentProperties segmentProperties =
+          new SegmentProperties(footerList.get(0).getColumnInTable(),
```

Comment:
I think the hashmap approach is OK. You can keep a hashmap in this class and store the mapping from a key (calculated from schema) to the SegmentProperties object.